File: backend/database/json_store.py
import json
import os
from threading import Lock

# Define the path for our JSON database file.
# It will be created in the same directory as this script.
DB_FILE = os.path.join(os.path.dirname(__file__), 'db.json')

# A lock to prevent race conditions when multiple requests try to write to the file.
file_lock = Lock()

def _initialize_db():
    """Ensures the database file exists and has a basic structure."""
    with file_lock:
        if not os.path.exists(DB_FILE):
            # Create the file with an empty structure for users
            with open(DB_FILE, 'w') as f:
                json.dump({"users": {}}, f, indent=4)

def read_data():
    """
    Reads the entire database from the JSON file in a thread-safe manner.
    """
    _initialize_db()  # Ensure file exists before reading
    with file_lock:
        with open(DB_FILE, 'r') as f:
            try:
                data = json.load(f)
            except json.JSONDecodeError:
                # If the file is empty or corrupted, return a default structure
                return {"users": {}}
    return data

def write_data(data: dict):
    """
    Writes the entire database to the JSON file in a thread-safe manner.
    """
    _initialize_db()  # Ensure file exists before writing
    with file_lock:
        with open(DB_FILE, 'w') as f:
            json.dump(data, f, indent=4)

# ==============================================================================
# User-specific data access functions
# ==============================================================================

def find_user_by_id(user_id: str) -> dict | None:
    """
    Finds a user by their ID (which will be the UID from Firebase).
    
    Args:
        user_id: The unique identifier for the user.
        
    Returns:
        The user's data as a dictionary, or None if not found.
    """
    db = read_data()
    return db.get("users", {}).get(user_id)

def save_user(user_data: dict):
    """
    Saves or updates a user's data in the JSON store.
    The user_data dictionary is expected to contain a 'uid' key.
    
    Args:
        user_data: A dictionary containing the user's details.
    """
    if 'uid' not in user_data:
        raise ValueError("User data must contain a 'uid' key.")
    
    user_id = user_data['uid']
    db = read_data()
    
    # Ensure the 'users' key exists
    if "users" not in db:
        db["users"] = {}
        
    db["users"][user_id] = user_data
    write_data(db)
    return user_data

# Initialize the database file on startup
_initialize_db()

File: backend/database/jsonStore.js
/**
 * JSON file-based storage for documents and users
 */
const fs = require('fs');
const path = require('path');

// Define paths
const DATA_DIR = path.join(__dirname, '..', '..', 'data');
const DB_FILE = path.join(__dirname, 'db.json');

// Ensure data directory exists
if (!fs.existsSync(DATA_DIR)) {
    fs.mkdirSync(DATA_DIR, { recursive: true });
}

// ==============================================================================
// Document Storage Functions (for OCR results and questions)
// ==============================================================================

/**
 * Get the file path for a document
 * @param {string} docId - Document ID
 * @returns {string} File path
 */
function getDocPath(docId) {
    return path.join(DATA_DIR, `${docId}.json`);
}

/**
 * Load a document by ID
 * @param {string} docId - Document ID
 * @returns {object|null} Document data or null if not found
 */
function loadDoc(docId) {
    const filePath = getDocPath(docId);
    if (!fs.existsSync(filePath)) {
        return null;
    }
    try {
        const content = fs.readFileSync(filePath, 'utf-8');
        return JSON.parse(content);
    } catch (error) {
        console.error(`Error loading document ${docId}:`, error);
        return null;
    }
}

/**
 * Save a document
 * @param {string} docId - Document ID
 * @param {object} data - Document data
 */
function saveDoc(docId, data) {
    const filePath = getDocPath(docId);
    fs.writeFileSync(filePath, JSON.stringify(data, null, 2), 'utf-8');
}

// ==============================================================================
// User Database Functions
// ==============================================================================

/**
 * Initialize the database file if it doesn't exist
 */
function initializeDb() {
    if (!fs.existsSync(DB_FILE)) {
        fs.writeFileSync(DB_FILE, JSON.stringify({ users: {} }, null, 4), 'utf-8');
    }
}

/**
 * Read the entire database
 * @returns {object} Database contents
 */
function readData() {
    initializeDb();
    try {
        const content = fs.readFileSync(DB_FILE, 'utf-8');
        return JSON.parse(content);
    } catch (error) {
        console.error('Error reading database:', error);
        return { users: {} };
    }
}

/**
 * Write the entire database
 * @param {object} data - Data to write
 */
function writeData(data) {
    initializeDb();
    fs.writeFileSync(DB_FILE, JSON.stringify(data, null, 4), 'utf-8');
}

/**
 * Find a user by their ID (Firebase UID)
 * @param {string} userId - User ID
 * @returns {object|null} User data or null if not found
 */
function findUserById(userId) {
    const db = readData();
    return db.users[userId] || null;
}

/**
 * Save or update a user
 * @param {object} userData - User data (must contain 'uid' key)
 * @returns {object} The saved user data
 */
function saveUser(userData) {
    if (!userData.uid) {
        throw new Error("User data must contain a 'uid' key.");
    }

    const userId = userData.uid;
    const db = readData();

    if (!db.users) {
        db.users = {};
    }

    db.users[userId] = userData;
    writeData(db);
    return userData;
}

// Initialize database on module load
initializeDb();

module.exports = {
    // Document functions
    DATA_DIR,
    getDocPath,
    loadDoc,
    saveDoc,

    // User functions
    findUserById,
    saveUser,
    readData,
    writeData
};
